ql/src/java/org/apache/hadoop/hive/ql/Driver.java

```diff
@@ -359,6 +216,101 @@ private void runInternal(String command, boolean alreadyCompiled) throws Command
+        // Inserts will not invalidate the snapshot, that could cause duplicates.
+        if (!driverTxnHandler.isValidTxnListState()) {
+          LOG.info("Re-compiling after acquiring locks, attempt #" + retryShapshotCount);
+          // Snapshot was outdated when locks were acquired, hence regenerate context, txn list and retry.
+          // TODO: Lock acquisition should be moved before analyze, this is a bit hackish.
+          // Currently, we acquire a snapshot, compile the query with that snapshot, and then - acquire locks.
+          // If snapshot is still valid, we continue as usual.
+          // But if snapshot is not valid, we recompile the query.
+          if (driverContext.isOutdatedTxn()) {
+            // Later transaction invalidated the snapshot, a new transaction is required
+            LOG.info("Snapshot is outdated, re-initiating transaction ...");
+            driverContext.getTxnManager().rollbackTxn();
+
+            String userFromUGI = DriverUtils.getUserFromUGI(driverContext);
+            driverContext.getTxnManager().openTxn(context, userFromUGI, driverContext.getTxnType());
+            lockAndRespond();
+          }
+          driverContext.setRetrial(true);
+          driverContext.getBackupContext().addSubContext(context);
+          driverContext.getBackupContext().setHiveLocks(context.getHiveLocks());
+          context = driverContext.getBackupContext();
+
+          driverContext.getConf().set(ValidTxnList.VALID_TXNS_KEY,
+              driverContext.getTxnManager().getValidTxns().toString());
+
+          if (driverContext.getPlan().hasAcidResourcesInQuery()) {
+            compileInternal(context.getCmd(), true);
+            driverTxnHandler.recordValidWriteIds();
+            driverTxnHandler.setWriteIdForAcidFileSinks();
+          }
+          // Since we're reusing the compiled plan, we need to update its start time for current run
+          driverContext.getPlan().setQueryStartTime(driverContext.getQueryDisplay().getQueryStartTime());
+        }
+        // Re-check snapshot only in case we had to release locks and open a new transaction,
+        // otherwise exclusive locks should protect output tables/partitions in snapshot from concurrent writes.
+      } while (driverContext.isOutdatedTxn() && ++retryShapshotCount <= maxRetrySnapshotCount);
+
+      if (retryShapshotCount > maxRetrySnapshotCount) {
+        // Throw exception
+        HiveException e = new HiveException(
+            "Operation could not be executed, " + SNAPSHOT_WAS_OUTDATED_WHEN_LOCKS_WERE_ACQUIRED + ".");
+        DriverUtils.handleHiveException(driverContext, e, 14, null);
+      } else if (retryShapshotCount != 0) {
+        // the reason that we set the txn manager for the cxt here is because each query has its own ctx object.
+        // The txn mgr is shared across the same instance of Driver, which can run multiple queries.
+        context.setHiveTxnManager(driverContext.getTxnManager());
+      }
```

Comment:
Since you are slicing and dicing things, you may want to look at breaking this out a bit.

It is bad form to throw and exception, and then handle it, within the same method.

I think this should be moved out of this `try` block.

```
      if (retryShapshotCount > maxRetrySnapshotCount) {
        // Throw exception
        HiveException e = new HiveException(
            "Operation could not be executed, " + SNAPSHOT_WAS_OUTDATED_WHEN_LOCKS_WERE_ACQUIRED + ".");
        DriverUtils.handleHiveException(driverContext, e, 14, null);
      }
  if (retryShapshotCount != 0) {
        // the reason that we set the txn manager for the cxt here is because each query has its own ctx object.
        // The txn mgr is shared across the same instance of Driver, which can run multiple queries.
        context.setHiveTxnManager(driverContext.getTxnManager());
        return true;
    }
    return false;
```

Reply:
It is probably overwriting the error code of '14' with the outer-try-catch code of '13'.

Reply:
DriverUtils.handleHiveException throws a CommandProcessorException which is not caught, as it is neither a LockException nor a SemanticException. Still the block that you've mentioned can be moved to a separate method, but I don't see the meaning of the returned boolean. What would that be used for?

Reply:
oh, I see the meaning of the return value, it is what needs to b returned from the original method. Sure, I can move this to a separate method, do you have a suggestion for the name?